l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,727 @@
+    /// finalizeDeposit uses the new format. On the L2 we have finalizeDeposit with new and old formats both.
+    function finalizeDeposit(FinalizeL1DepositParams calldata _finalizeWithdrawalParams) public {
+        _finalizeDeposit(_finalizeWithdrawalParams);
+    }
+
+    /// @notice Internal function that handles the logic for finalizing withdrawals, supporting both the current bridge system and the legacy ERC20 bridge.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    function _finalizeDeposit(
+        FinalizeL1DepositParams calldata _finalizeWithdrawalParams
+    ) internal nonReentrant whenNotPaused {
+        // returns (address l1Receiver, address l1Token, uint256 amount) {
+        bytes memory transferData;
+        bytes32 assetId;
+        (assetId, transferData) = _verifyAndGetWithdrawalData(_finalizeWithdrawalParams);
+        l1AssetRouter.finalizeDeposit(_finalizeWithdrawalParams.chainId, assetId, transferData);
+        // if ((l1NativeTokenVault.tokenAddress(assetId) != address(0)) && (!l1NativeTokenVault.isTokenBridged(assetId))){
+        //     l1Receiver = address(l1NativeTokenVault);
+        //     l1Token = address(0);
+        //     amount = transferData[1];
+        //     (l1Receiver, amount) = DataEncoding.decodeTransferData(transferData);
+        // }
+    }
+
+    /// @notice Internal function that handles the logic for finalizing withdrawals, supporting both the current bridge system and the legacy ERC20 bridge.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    /// @return assetId The bridged asset ID.
+    /// @return transferData The encoded transfer data.
+    function _verifyAndGetWithdrawalData(
+        FinalizeL1DepositParams calldata _finalizeWithdrawalParams
+    ) internal whenNotPaused returns (bytes32 assetId, bytes memory transferData) {
+        if (
+            isWithdrawalFinalized[_finalizeWithdrawalParams.chainId][_finalizeWithdrawalParams.l2BatchNumber][
+                _finalizeWithdrawalParams.l2MessageIndex
+            ]
+        ) {
+            revert WithdrawalAlreadyFinalized();
+        }
+        isWithdrawalFinalized[_finalizeWithdrawalParams.chainId][_finalizeWithdrawalParams.l2BatchNumber][
+            _finalizeWithdrawalParams.l2MessageIndex
+        ] = true;
+
+        // Handling special case for withdrawal from ZKsync Era initiated before Shared Bridge.
+        require(
+            !_isEraLegacyEthWithdrawal(_finalizeWithdrawalParams.chainId, _finalizeWithdrawalParams.l2BatchNumber),
+            "L1N: legacy eth withdrawal"
+        );
+        require(
+            !_isEraLegacyTokenWithdrawal(_finalizeWithdrawalParams.chainId, _finalizeWithdrawalParams.l2BatchNumber),
+            "L1N: legacy token withdrawal"
+        );
+
+        (assetId, transferData) = _checkWithdrawal(_finalizeWithdrawalParams);
```

Comment:
nit: `_verifyAndClearFailedTransfer` but `_checkWithdrawal`, we should use either `check` or `verify` but not both to not confuse for different purposes

Reply:
of renaming _checkWithdrawal to _verifyWithdrawal